### PR TITLE
test: use more frequent reconnection policy

### LINKED ISSUE #15104: topology tests: too low timeouts used for consistency checks
For example test/topology/test_replace.py:
```python3
@pytest.mark.asyncio
async def test_replace_different_ip(manager: ManagerClient) -> None:
    """Replace an existing node with new node using a different IP address"""
    servers = await manager.running_servers()
    await manager.server_stop(servers[0].server_id)
    replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = False, use_host_id = False)
    await manager.server_add(replace_cfg)
    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
```

The timeout used for `wait_for_token_ring_and_group0_consistency` is 30 seconds.

Turns out it's too low, because for some reason, the driver sometimes sleeps for ~30 seconds before connecting to a node, and we need the driver connected to fetch group 0 member list.

(And they said that using CQL instead of REST APIs was a good idea.)

Example logs from CI run: 
[topology.test_replace.3.log](https://github.com/scylladb/scylladb/files/12394336/topology.test_replace.3.log)

```
22:47:46.054 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:47.056 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:48.059 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:49.061 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:50.063 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:51.065 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:52.067 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:53.068 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:54.069 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:55.071 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:56.073 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:57.075 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:58.076 INFO> Driver not connected to 127.204.54.40:9042 yet
22:47:59.077 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:00.079 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:01.080 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:02.082 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:03.083 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:04.085 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:05.086 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:06.088 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:07.090 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:08.092 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:09.093 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:10.094 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:11.095 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:12.097 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:13.098 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:14.100 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:15.102 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:16.104 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:17.105 INFO> Driver not connected to 127.204.54.40:9042 yet
22:48:17.913 DEBUG> Connection 140694068603152 127.0.0.1:39308 -> 127.204.54.40:9042
```

cc @avelanarius 

what we can do in the tests is increase the timeouts...

### PR DESCRIPTION
**Pull request by avelanarius**
The default reconnection policy in Python Driver is an exponential backoff (with jitter) policy, which starts at 1 second reconnection interval and ramps up to 600 seconds.

This is a problem in tests (refs #15104), especially in tests that restart or replace nodes. In such a scenario, a node can be unavailable for an  extended period of time and the driver will try to reconnect to it  multiple times, eventually reaching very long reconnection interval  values, exceeding the timeout of a test.

Fix the issue by using an exponential reconnection policy with a maximum interval of 4 seconds. A smaller value was not chosen, as each retry clutters the logs with reconnection exception stack trace.

Fixes #15104

### CODE REVIEW

File: test/cql-pytest/run.py [hunk 1]
```diff
@@ -344,7 +344,13 @@ def get_cql_cluster(ip, ssl_context=None):
...
         # so let's increase them to 60 seconds. See issue #13239.
         connect_timeout = 60,
-        control_connection_timeout = 60)
+        control_connection_timeout = 60,
+        # The default reconnection policy has a large maximum interval
+        # between retries (600 seconds). In tests that restart/replace nodes,
+        # where a node can be unavailable for an extended period of time,
+        # this can cause the reconnection retry interval to get very large,
+        # longer than a test timeout.
+        reconnection_policy = cassandra.policies.ExponentialReconnectionPolicy(1.0, 4.0))
```
Comment (nyh) [inline test/cql-pytest/run.py:353]: What you are changing here is only the connection that run.py uses to check that Scylla came up. The code that is used by actual cql-pytest tests is cql_session() in test/cql-pytest/util.py. 

I don't think this change is important to cql-pytest anyway (which doesn't need to reconnect, Scylla is always assumed to be alive) - so you can drop it from here and not add it to util.py

File: test/cql-pytest/test_ssl.py [hunk 2]
```diff
@@ -62,7 +62,13 @@ def try_connect(orig_cluster, ssl_version):
...
         # so let's increase them to 60 seconds. See issue #11289.
         connect_timeout = 60,
-        control_connection_timeout = 60)
+        control_connection_timeout = 60,
+        # The default reconnection policy has a large maximum interval
+        # between retries (600 seconds). In tests that restart/replace nodes,
+        # where a node can be unavailable for an extended period of time,
+        # this can cause the reconnection retry interval to get very large,
+        # longer than a test timeout.
+        reconnection_policy = cassandra.policies.ExponentialReconnectionPolicy(1.0, 4.0))
```
Comment (nyh) [inline test/cql-pytest/test_ssl.py:71]: Again, this isn't relevant to this specific test, which never needs to reconnect.

File: test/cql-pytest/util.py [hunk 3]
```diff
@@ -191,6 +191,12 @@ def cql_session(host, port, is_ssl, username, password):
         # so let's increase them to 60 seconds. See issue #11289.
         connect_timeout = 60,
         control_connection_timeout = 60,
+        # The default reconnection policy has a large maximum interval
+        # between retries (600 seconds). In tests that restart/replace nodes,
+        # where a node can be unavailable for an extended period of time,
+        # this can cause the reconnection retry interval to get very large,
+        # longer than a test timeout.
+        reconnection_policy = ExponentialReconnectionPolicy(1.0, 4.0)
```
Comment (nyh) [inline test/cql-pytest/util.py:199]: As I noted above this change shouldn't really help cql-pytest. And the comment suggests it does something which doesn't happen in cql-pytest.
I'm not strongly against this line (here, not in the other files where it's completely out of place), but I think you should consider if it's really useful, and if not, drop it.

File: test/rest_api/conftest.py [hunk 4]
```diff
@@ -100,6 +100,12 @@ def cql(request):
         # so let's increase them to 60 seconds. See issue #11289.
         connect_timeout = 60,
         control_connection_timeout = 60,
+        # The default reconnection policy has a large maximum interval
+        # between retries (600 seconds). In tests that restart/replace nodes,
+        # where a node can be unavailable for an extended period of time,
+        # this can cause the reconnection retry interval to get very large,
+        # longer than a test timeout.
+        reconnection_policy = ExponentialReconnectionPolicy(1.0, 4.0)
```
Comment (nyh) [inline test/rest_api/conftest.py:108]: Again, is this change relevant to the rest_api tests?

**Review by nyh — CHANGES_REQUESTED** [review]

File: test/cql-pytest/run.py (hunk 1 above)
Comment (avelanarius) [reply test/cql-pytest/run.py:353]: My reasons for copying those parameters to all places were:
- Almost all `Cluster` constructors in `test/` use the same options, so I wanted it to remain consistent
- Setting this particular reconnection policy is non-pessimizing and it's a better default for tests in general
- I wanted to avoid a similar situation as was with `connect_timeout` and `control_connection_timeout`, where those options were originally not added to all places and 4 separate commits at separate times eventually propagated it to all places: 889acf710cac2cc464c74926df84946fe8086c98, d408b711e3f666448412ca7155c3c24360435b77, 4fdcee8415b1b797b9c69ac159a98390199ca8a5, 055340ae39ba7701e798ca370d01d343c28ac231. However, I now see that `connect_timeout` and `control_connection_timeout` are really needed in all tests (for debug build of Scylla), while `reconnection_policy` isn't needed everywhere.

**Comment by avelanarius** [comment]
v2:
- Removed changing `reconnection_policy` in tests that don't do any topology changes. The only changed file is now `test/topology/conftest.py`

**Review by kbr-scylla — APPROVED** [review]

**Review by nyh — APPROVED** [review]

**Comment by bhalevy** [comment]
Looking at https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3184/artifact/testlog/x86_64/debug/topology.test_replace.1.log

I'm not sure that this fix will be sufficient.

```
19:23:25.431 INFO> Group 0 members by Server(670, 127.231.175.33): {('c78c3712-6b92-4526-ba85-ed9deabb2549', True), ('f9ad94bc-f455-474f-9610-53b33308d51c', True), ('1ed252e5-ac55-4e32-b236-ca3ea87ade8f', True)}
19:23:25.431 DEBUG> RESTClient fetching GET http://127.231.175.33:10000/storage_service/tokens_endpoint
19:23:25.465 INFO> Normal endpoints' IPs by Server(670, 127.231.175.33): {'127.231.175.33', '127.231.175.8', '127.231.175.17'}
19:23:25.465 DEBUG> RESTClient fetching GET http://127.231.175.33:10000/storage_service/host_id
19:23:25.479 INFO> All host IDs by Server(670, 127.231.175.33): {'1ed252e5-ac55-4e32-b236-ca3ea87ade8f', 'c78c3712-6b92-4526-ba85-ed9deabb2549', 'f9ad94bc-f455-474f-9610-53b33308d51c'}
19:23:25.479 INFO> Normal endpoints' host IDs by Server(670, 127.231.175.33): {'1ed252e5-ac55-4e32-b236-ca3ea87ade8f', 'c78c3712-6b92-4526-ba85-ed9deabb2549', 'f9ad94bc-f455-474f-9610-53b33308d51c'}
19:23:25.551 INFO> Group 0 members by Server(690, 127.231.175.17): {('c78c3712-6b92-4526-ba85-ed9deabb2549', True), ('f9ad94bc-f455-474f-9610-53b33308d51c', True), ('1ed252e5-ac55-4e32-b236-ca3ea87ade8f', True)}
19:23:25.552 DEBUG> RESTClient fetching GET http://127.231.175.17:10000/storage_service/tokens_endpoint
19:23:25.573 INFO> Normal endpoints' IPs by Server(690, 127.231.175.17): {'127.231.175.33', '127.231.175.8', '127.231.175.17'}
19:23:25.573 DEBUG> RESTClient fetching GET http://127.231.175.17:10000/storage_service/host_id
19:23:25.583 INFO> All host IDs by Server(690, 127.231.175.17): {'1ed252e5-ac55-4e32-b236-ca3ea87ade8f', 'c78c3712-6b92-4526-ba85-ed9deabb2549', 'f9ad94bc-f455-474f-9610-53b33308d51c'}
19:23:25.583 INFO> Normal endpoints' host IDs by Server(690, 127.231.175.17): {'1ed252e5-ac55-4e32-b236-ca3ea87ade8f', 'c78c3712-6b92-4526-ba85-ed9deabb2549', 'f9ad94bc-f455-474f-9610-53b33308d51c'}
19:23:25.584 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:26.586 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:27.587 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:28.589 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:29.591 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:30.593 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:31.595 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:32.596 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:33.598 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:34.599 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:35.601 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:36.603 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:37.604 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:38.606 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:39.607 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:40.609 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:41.611 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:42.611 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:43.612 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:44.614 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:45.616 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:46.618 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:47.619 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:48.620 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:49.622 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:50.624 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:51.625 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:52.627 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:53.628 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:54.629 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:55.629 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:56.631 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:57.633 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:58.634 INFO> Driver not connected to 127.231.175.8:9042 yet
19:23:59.512 DEBUG> Sending options message heartbeat on idle connection (140368998624720) 127.231.175.17:9042
19:23:59.513 DEBUG> Sending options message heartbeat on idle connection (140368998264208) 127.231.175.17:9042
19:23:59.513 DEBUG> Sending options message heartbeat on idle connection (140368996278800) 127.231.175.33:9042
19:23:59.513 DEBUG> Sending options message heartbeat on idle connection (140368998220112) 127.231.175.33:9042
19:23:59.514 DEBUG> Sending options message heartbeat on idle connection (140368998595152) 127.231.175.17:9042
19:23:59.517 DEBUG> Received options response on connection (140368998595152) from 127.231.175.17:9042
19:23:59.518 DEBUG> Received options response on connection (140368998624720) from 127.231.175.17:9042
19:23:59.518 DEBUG> Received options response on connection (140368998264208) from 127.231.175.17:9042
19:23:59.519 DEBUG> Received options response on connection (140368998220112) from 127.231.175.33:9042
19:23:59.519 DEBUG> Received options response on connection (140368996278800) from 127.231.175.33:9042
19:23:59.636 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:00.638 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:01.640 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:02.642 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:03.643 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:04.644 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:05.645 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:06.646 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:07.647 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:08.648 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:09.650 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:10.651 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:11.653 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:12.654 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:13.655 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:14.657 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:15.658 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:16.660 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:17.662 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:18.664 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:19.666 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:20.666 INFO> Driver not connected to 127.231.175.8:9042 yet
19:24:21.021 DEBUG> Connection 140368998078288 127.0.0.1:36462 -> 127.231.175.8:9042
19:24:21.021 DEBUG> Sending initial options message for new connection (140368998078288) to 127.231.175.8:9042
19:24:21.024 DEBUG> Received options response on new connection (140368998078288) from 127.231.175.8:9042
19:24:21.025 DEBUG> No available compression types supported on both ends. locally supported: odict_keys([]). remotely supported: ['lz4', 'snappy']
19:24:21.025 DEBUG> Sending StartupMessage on <LibevConnection(140368998078288) 127.231.175.8:9042>
19:24:21.025 DEBUG> Sent StartupMessage on <LibevConnection(140368998078288) 127.231.175.8:9042>
19:24:21.028 DEBUG> Got ReadyMessage on new connection (140368998078288) from 127.231.175.8:9042
19:24:21.028 INFO> Successful reconnection to 127.231.175.8:9042, marking node up if it isn't already
19:24:21.029 DEBUG> Waiting to acquire lock for handling up status of node 127.231.175.8:9042
19:24:21.029 DEBUG> Starting to handle up status of node 127.231.175.8:9042
19:24:21.029 INFO> Host 127.231.175.8:9042 may be up; will prepare queries and open connection pool
19:24:21.029 DEBUG> Now that host 127.231.175.8:9042 is up, cancelling the reconnection handler
19:24:21.030 DEBUG> Done preparing all queries for host 127.231.175.8:9042, 
19:24:21.030 DEBUG> Signalling to load balancing policies that host 127.231.175.8:9042 is up
19:24:21.030 DEBUG> Signalling to control connection that host 127.231.175.8:9042 is up
19:24:21.031 DEBUG> Attempting to open new connection pools for host 127.231.175.8:9042
19:24:21.031 DEBUG> Closing connection (140368998078288) to 127.231.175.8:9042
19:24:21.031 DEBUG> Closed socket to 127.231.175.8:9042
19:24:21.032 DEBUG> Initializing connection for host 127.231.175.8:9042
19:24:21.033 DEBUG> Connection 140368998079376 127.0.0.1:36468 -> 127.231.175.8:9042
19:24:21.033 DEBUG> Sending initial options message for new connection (140368998079376) to 127.231.175.8:9042
19:24:21.035 DEBUG> Received options response on new connection (140368998079376) from 127.231.175.8:9042
19:24:21.036 DEBUG> No available compression types supported on both ends. locally supported: odict_keys([]). remotely supported: ['lz4', 'snappy']
19:24:21.036 DEBUG> Sending StartupMessage on <LibevConnection(140368998079376) 127.231.175.8:9042>
19:24:21.036 DEBUG> Sent StartupMessage on <LibevConnection(140368998079376) 127.231.175.8:9042>
19:24:21.038 DEBUG> Got ReadyMessage on new connection (140368998079376) from 127.231.175.8:9042
19:24:21.038 DEBUG> First connection created to 127.231.175.8:9042 for shard_id=0
19:24:21.039 DEBUG> Finished initializing connection for host 127.231.175.8:9042
19:24:21.039 DEBUG> Added pool for host 127.231.175.8:9042 to session
19:24:21.039 INFO> Connection pools established for node 127.231.175.8:9042
19:24:21.039 DEBUG> Host 127.231.175.8:9042 is now marked up
19:24:21.039 DEBUG> shard_aware_endpoint=None
19:24:21.040 DEBUG> Connection 140368998489424 127.0.0.1:36476 -> 127.231.175.8:9042
19:24:21.040 DEBUG> Sending initial options message for new connection (140368998489424) to 127.231.175.8:9042
19:24:21.044 DEBUG> Received options response on new connection (140368998489424) from 127.231.175.8:9042
19:24:21.045 DEBUG> No available compression types supported on both ends. locally supported: odict_keys([]). remotely supported: ['lz4', 'snappy']
19:24:21.045 DEBUG> Sending StartupMessage on <LibevConnection(140368998489424) 127.231.175.8:9042>
19:24:21.045 DEBUG> Sent StartupMessage on <LibevConnection(140368998489424) 127.231.175.8:9042>
19:24:21.049 DEBUG> Got ReadyMessage on new connection (140368998489424) from 127.231.175.8:9042
19:24:21.049 DEBUG> Received a connection 140368998489424 for shard_id=0 on host 127.231.175.8:9042
19:24:21.049 DEBUG> Putting a connection 140368998489424 to shard 0 to the excess pool of host 127.231.175.8:9042
19:24:21.753 INFO> Group 0 members by Server(734, 127.231.175.8): {('c78c3712-6b92-4526-ba85-ed9deabb2549', True), ('f9ad94bc-f455-474f-9610-53b33308d51c', True), ('1ed252e5-ac55-4e32-b236-ca3ea87ade8f', True)}
FAILED
...
_____________________ test_replace_reuse_ip_using_host_id ______________________

manager = <test.pylib.manager_client.ManagerClient object at 0x7faa36732a90>

    @pytest.mark.asyncio
    async def test_replace_reuse_ip_using_host_id(manager: ManagerClient) -> None:
        """Replace an existing node with new node using the same IP address and same host id"""
        servers = await manager.running_servers()
        await manager.server_stop(servers[0].server_id)
        replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = True, use_host_id = True)
        await manager.server_add(replace_cfg)
>       await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)

test/topology/test_replace.py:50: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/topology/util.py:94: in wait_for_token_ring_and_group0_consistency
    await wait_for(token_ring_matches, deadline, period=.5)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pred = <function wait_for_token_ring_and_group0_consistency.<locals>.token_ring_matches at 0x7faa343eae80>
deadline = 1692721435.3082569, period = 0.5

    async def wait_for(
            pred: Callable[[], Awaitable[Optional[T]]],
            deadline: float, period: float = 1) -> T:
        while True:
>           assert(time.time() < deadline), "Deadline exceeded, failing test."
E           AssertionError: Deadline exceeded, failing test.
```

The deadline, 1692721435, is 2023-08-22T16:23:55 UTC

But the node started listening to CQL much earlier, and the driver still wasn't able to connect for some reason.
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3184/artifact/testlog/x86_64/debug/scylla-734.log
```
INFO  2023-08-22 19:22:13,464 [shard 0] storage_service - Host 1ed252e5-ac55-4e32-b236-ca3ea87ade8f/127.231.175.8 is replacing 9642ca85-083f-44ec-b01d-a1ac8ba19903/127.231.175.8
...
INFO  2023-08-22 19:23:23,961 [shard 0] cql_server_controller - Starting listening for CQL clients on 127.231.175.8:9042 (unencrypted, non-shard-aware)
INFO  2023-08-22 19:23:23,962 [shard 0] cql_server_controller - Starting listening for CQL clients on 127.231.175.8:19042 (unencrypted, shard-aware)
INFO  2023-08-22 19:23:23,995 [shard 0] init - serving
INFO  2023-08-22 19:23:23,995 [shard 0] init - Scylla version 5.4.0~dev-0.20230822.d1966ba8986a initialization completed.
INFO  2023-08-22 19:23:25,291 [shard 0] compaction - [Compact system.scylla_local 362f2ff0-4108-11ee-9685-a7371d4d7cc8] Compacted 2 sstables to [/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/scylla_local-2972ec7ffb2038ddaac1d876f2e3fcbd/me-3g8q_19iy_0f7u82aeklz6l2c5hk-big-Data.db:level=0]. 65kB to 59kB (~90% of original) in 2820ms = 23kB/s. ~256 total partitions merged to 2.
INFO  2023-08-22 19:23:25,298 [shard 0] compaction - [Compact system.local 382ad8e0-4108-11ee-9685-a7371d4d7cc8] Compacting [/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/local-7ad54392bcdd35a684174e047860b377/me-3g8q_19iz_2j3bk2aeklz6l2c5hk-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/local-7ad54392bcdd35a684174e047860b377/me-3g8q_19iz_0n5c02aeklz6l2c5hk-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/local-7ad54392bcdd35a684174e047860b377/me-3g8q_19iw_0r0802aeklz6l2c5hk-big-Data.db:level=0:origin=compaction]
INFO  2023-08-22 19:23:25,535 [shard 0] compaction - [Compact system.local 382ad8e0-4108-11ee-9685-a7371d4d7cc8] Compacted 3 sstables to [/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/local-7ad54392bcdd35a684174e047860b377/me-3g8q_19j1_1yqao2aeklz6l2c5hk-big-Data.db:level=0]. 21kB to 8kB (~38% of original) in 168ms = 125kB/s. ~384 total partitions merged to 1.
INFO  2023-08-22 19:23:25,543 [shard 0] compaction - [Compact system.peers 38501420-4108-11ee-9685-a7371d4d7cc8] Compacting [/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-3g8q_19ix_112vk2aeklz6l2c5hk-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-3g8q_19iy_00ffk2aeklz6l2c5hk-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-3g8q_19iy_1ekzk2aeklz6l2c5hk-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-3g8q_19iw_25l7k2aeklz6l2c5hk-big-Data.db:level=0:origin=compaction]
INFO  2023-08-22 19:23:26,448 [shard 0] compaction - [Compact system.peers 38501420-4108-11ee-9685-a7371d4d7cc8] Compacted 4 sstables to [/jenkins/workspace/scylla-master/scylla-ci/scylla/testlog/x86_64/debug/scylla-734/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-3g8q_19j1_3fgg02aeklz6l2c5hk-big-Data.db:level=0]. 199kB to 63kB (~31% of original) in 816ms = 243kB/s. ~512 total partitions merged to 2.
INFO  2023-08-22 19:24:21,716 [shard 0] gossip - 60000 ms elapsed, 127.231.175.8 gossip quarantine over
------ Ending test debug/topology.test_replace.1::test_replace_reuse_ip_using_host_id ------
```

Having per-server timeouts as in https://github.com/scylladb/scylladb/pull/14983 doesn't seem like a bad idea after all :)

**Comment by kbr-scylla** [comment]
> But the node started listening to CQL much earlier, and the driver still wasn't able to connect for some reason.

It's not that it wasn't able to connect, it didn't even try (due to the long sleep due to the exponential retry policy).

Before the node starts listening:
```
19:23:22.095 WARNING> Error attempting to reconnect to 127.231.175.8:9042, scheduling retry in 58.88 seconds: [Errno 111] Tried connecting to [('127.231.175.8', 9042)]. Last error: Connection refused
```

the driver schedules the next retry to happen **a minute later**... and that was ~2 seconds before the node started listening:
```
INFO  2023-08-22 19:23:23,961 [shard 0] cql_server_controller - Starting listening for CQL clients on 127.231.175.8:9042 (unencrypted, non-shard-aware)
INFO  2023-08-22 19:23:23,962 [shard 0] cql_server_controller - Starting listening for CQL clients on 127.231.175.8:19042 (unencrypted, shard-aware)
```

So now the driver would sleep for a minute doing nothing, even if it could have already connected in 2 seconds after the last attempt.

That is precisely the problem that this PR is solving.

**Comment by bhalevy** [comment]
> the driver schedules the next retry to happen a minute later... and that was ~2 seconds before the node started listening:

Sorry, I missed that.
Thanks for verifying.
